### PR TITLE
extracting a subset of the Pareto front

### PR DESCRIPTION
**Pull request by teytaud**
3 methods for extracting a subset of a Pareto set:
(a) random sampling
(b) best covering in the objective functions space
(c) best covering in the domain

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by teytaud** [comment]
one day i will find how I can add someone like @brockho  as a reviewer....

File: nevergrad/functions/multiobjective/core.py [hunk 1]
```diff
@@ -74,8 +75,7 @@ def __call__(self, *args: Any, **kwargs: Any) -> float:
         # The following is not. It should be called locally.
         return self.compute_aggregate_loss(losses, *args, **kwargs)
 
-    @property
-    def pareto_front(self) -> List[ArgsKwargs]:
+    def filter_pareto_front(self):
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: You should probably make this private: `_filter_pareto_front`

File: nevergrad/functions/multiobjective/core.py [hunk 2]
```diff
@@ -89,4 +89,34 @@ def pareto_front(self) -> List[ArgsKwargs]:
             if should_be_added:
                 new_points.append((argskwargs, losses))
         self._points = new_points
-        return [p[0] for p in self._points]
+
+    def pareto_front(self, size: Optional[int] = None, method: Optional[str] = None) -> List[ArgsKwargs]:
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: I think yo add a better name: `subset` seems clearer than `method`

File: nevergrad/functions/multiobjective/core.py [hunk 3]
```diff
@@ -89,4 +89,34 @@ def pareto_front(self) -> List[ArgsKwargs]:
...
+        """Pareto front, as a list of args and kwargs (tuple of a tuple and a dict)
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: With the new documentation, we'll need to level up the docstrings :D :
```
"""Pareto front, as a list of args and kwargs (tuple of a tuple and a dict)

Parameters
------------
size:  int (optional)
    if provided, selects a subset of the full pareto front with the given maximum size
subset: str
    method for selecting the subset ("random, "loss-covering", "domain-covering")

Returns
--------
list
    the list of elements of the pareto front
"""
```

File: nevergrad/functions/multiobjective/core.py (hunk 2 above)
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]:
```suggestion
    def pareto_front(self, size: Optional[int] = None, subset: str = random) -> List[ArgsKwargs]:
```

File: nevergrad/functions/multiobjective/core.py [hunk 4]
```diff
@@ -89,4 +89,34 @@ def pareto_front(self) -> List[ArgsKwargs]:
...
+        for the function - restricted to number elements covering as well as possible the domain 
+        (domain-covering) or the losses (loss-covering) or just randomly (random)..
+        """
+        self.filter_pareto_front()
+        if method is None:
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: You can remove this, let's give a default value to the parameter instead

Comment (jrapin) [reply nevergrad/functions/multiobjective/core.py]: Btw, what is "random" the best default?

File: nevergrad/functions/multiobjective/core.py [hunk 5]
```diff
@@ -89,4 +89,34 @@ def pareto_front(self) -> List[ArgsKwargs]:
...
+                    if method == "loss-covering":
+                        best_score = min(best_score, np.linalg.norm(ploss - vloss))
+                    elif method == "domain-covering":
+                        best_score = min(best_score, np.linalg.norm(tuple(i - j for i, j in zip(p[0][0], v[0][0]))))
+                    else:
+                        raise ValueError(f'Unknown method for Pareto-Set subsampling: "{method}"')
+                score += best_score ** 2
+            scores += [score]
+        return [p[0] for p in possibilities[scores.index(min(scores))]]
+        #return [p[0] for p in possibilities[scores.index(min(scores))]]
```
Comment (jrapin) [inline nevergrad/functions/multiobjective/core.py]: remove comment

**Comment by jrapin** [comment]
Also, a test would be nice :D

File: nevergrad/functions/multiobjective/core.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: Done.

File: nevergrad/functions/multiobjective/core.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: Done.

File: nevergrad/functions/multiobjective/core.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: Done.

File: nevergrad/functions/multiobjective/core.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: ah already did it, thx for the effort.

File: nevergrad/functions/multiobjective/core.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: Not sure it's the best... it's the fastest...

File: nevergrad/functions/multiobjective/core.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/functions/multiobjective/core.py]: Done.

**Comment by teytaud** [comment]
test added

done in a rush while coding in polygames with Maastricht guys, so not sure of what I am doing :-)

**Comment by brockho** [comment]
> 
> 
> one day i will find how I can add someone like @brockho as a reviewer....

I also don't know why this does not work. According to [the documentation](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews), everybody with read access (which I have) should be able to become reviewer. Maybe, you have to add me to the organization ("nevergrad"?) in order to see me in the list of potential reviewers.

**Pull request merged**